include/net/dsa.h

```diff
@@ -0,0 +1,283 @@
+ * @param flags Flags to be set in the entry
+ *
+ * @return 0 if ok, < 0 if error
+ */
+int dsa_set_mac_table_entry(const uint8_t *mac, uint8_t fw_port,
+			    uint16_t tbl_entry_idx, uint16_t flags);
+
+/**
+ * @brief Get DSA MAC address table entry
+ *
+ * @param buf The buffer for data read from the table
+ * @param tbl_entry_idx The index of entry in the table
+ *
+ * @return 0 if ok, < 0 if error
+ */
+int dsa_get_mac_table_entry(uint8_t *buf, uint16_t tbl_entry_idx);
+
+/**
+ * @brief Structure to provide dsa context
+ */
+
+struct dsa_context {
+	uint8_t num_slave_ports;
+	struct k_delayed_work dsa_work;
+	struct net_if *iface_slave[NET_DSA_PORT_MAX_COUNT];
+	struct net_if *iface_master;
+	bool link_up[NET_DSA_PORT_MAX_COUNT];
+
+	int (*sw_read)(uint16_t reg_addr, uint8_t *value);
+	int (*sw_write)(uint16_t reg_addr, uint8_t value);
+
+	uint8_t mac_addr[NET_DSA_PORT_MAX_COUNT][6];
+};
+
+#ifdef __cplusplus
+}
+#endif
+
+/* DSA Macros */
+#define DSA_IS_PORT_MASTER(iface) \
+	(net_eth_get_hw_capabilities(iface) & ETHERNET_DSA_MASTER_PORT)
+
+/* Device tree related, generic macros */
+#define DSA_DT_NODE DT_CHILD(DT_ALIAS(dsa_spi), dsa_0)
+#define DSA_DT_SLAVE_PORT_NUM \
+	DT_PROP(DSA_DT_NODE, dsa_slave_ports)
+#define DSA_DT_MASTER_PORT_NODE \
+	DT_PHANDLE(DSA_DT_NODE, dsa_master_port)
+#define DSA_DT_SPI_CPHA DT_PROP(DSA_DT_NODE, spi_cpha)
+#define DSA_DT_SPI_CPOL DT_PROP(DSA_DT_NODE, spi_cpol)
+
+extern const struct ethernet_api dsa_api_funcs;
+int dsa_port_init(struct device *dev);
+
+/*
+ * The order of NET_DEVICE_INIT_INSTANCE() placement IS important.
+ *
+ * To make the code simpler - the special care needs to be put on
+ * the proper placement of eth0, lan1, lan2, lan3, etc - to avoid
+ * the needs to search for proper interface when each packet is
+ * received or sent.
+ * The net_if.c has a very fast API to provide access to linked by
+ * the linker struct net_if(s) via device or index. As it is already
+ * available for use - let's use it.
+ *
+ * To do that one needs to check how linker places the interfaces.
+ * To inspect:
+ * objdump -dst ./zephyr/CMakeFiles/zephyr.dir/drivers/ethernet/eth_mcux.c.obj\
+ * | grep "__net_if"
+ * (The real problem is with eth0 and lanX order)
+ *
+ * If this approach is not enough for a simple system (like e.g. ip_k66f, one
+ * can prepare dedicated linker script for the board to force the
+ * order for complicated designs (like ones with eth0, eth1, and lanX).
+ *
+ * For simple cases it is just good enough.
+ */
+
+#if DT_NODE_HAS_PROP(DSA_DT_NODE, local_mac_address_port1)
+#define DSA_PORT1_MAC DT_PROP(DSA_DT_NODE, local_mac_address_port1)
+#endif
+#if DT_NODE_HAS_PROP(DSA_DT_NODE, local_mac_address_port2)
+#define DSA_PORT2_MAC DT_PROP(DSA_DT_NODE, local_mac_address_port2)
+#endif
+#if DT_NODE_HAS_PROP(DSA_DT_NODE, local_mac_address_port3)
+#define DSA_PORT3_MAC DT_PROP(DSA_DT_NODE, local_mac_address_port3)
+#endif
+
+#define DSA_NET_SLAVE_PORTS_INIT()	  \
```

Comment:
This is now mcux specific, should we place this to mcux driver header file in `drivers/ethernet/` directory instead?

Reply:
This `#define DSA_NET_SLAVE_PORTS_INIT()` uses generic `NET_DEVICE_INIT_INSTANCE()` to instantiate network interfaces. Other functions - like reading byte from DSA IC looks to be generic. Hence the idea to store it in dsa.h.

Reply:
Sure, but the macro uses MCUX specific settings. One option is to make them as parameters to this macro.
For example, the instance of the context variable could be a parameter to this macro, same for ETH_MCUX_PM_FUNC .

Reply:
Or I can remove them and just use the MAC address assigned for the eth0 (master) interface.

Reply:
Please find updated PR code. It now is more clear and has been moved to dsa_ksz8794.c driver file.